flow-typed/react-flip-move_v2.9.x.js

```diff
@@ -30,56 +28,80 @@ declare module 'react-flip-move' {
   declare export type ElementShape = {
-    type: $PropertyType<Element<*>, 'type'>,
-    props: $PropertyType<Element<*>, 'props'>,
-    key: $PropertyType<Element<*>, 'key'>,
-    ref: $PropertyType<Element<*>, 'ref'>,
+    +type: $PropertyType<React$Element<*>, "type">,
+    +props: $PropertyType<React$Element<*>, "props">,
+    +key: $PropertyType<React$Element<*>, "key">,
+    +ref: $PropertyType<React$Element<*>, "ref">
   };
 
   declare type ChildHook = (element: ElementShape, node: ?HTMLElement) => mixed;
 
   declare export type ChildrenHook = (
-    elements: Array<Element<*>>,
+    elements: Array<ElementShape>,
     nodes: Array<?HTMLElement>
   ) => mixed;
 
   declare export type GetPosition = (node: HTMLElement) => ClientRect;
 
-  declare export type VerticalAlignment = 'top' | 'bottom';
+  declare export type VerticalAlignment = "top" | "bottom";
+
+  // same as React.Node, but without fragments, see https://github.com/facebook/flow/issues/4781
+  declare export type Child = void | null | boolean | React$Element<*>;
+
+  // can't import from React, see https://github.com/facebook/flow/issues/4787
+  declare type ChildrenArray<T> = $ReadOnlyArray<ChildrenArray<T>> | T;
 
-  declare export type CommonProps = {
-    children: Node,
+  declare type BaseProps = {
     easing: string,
     typeName: string,
     disableAllAnimations: boolean,
     getPosition: GetPosition,
     maintainContainerHeight: boolean,
-    verticalAlignment: VerticalAlignment,
-    onStart?: ChildHook,
-    onFinish?: ChildHook,
-    onStartAll?: ChildrenHook,
-    onFinishAll?: ChildrenHook,
+    verticalAlignment: VerticalAlignment
   };
 
-  declare export type DelegatedProps = {
-    style?: ReactStyles,
-  };
-
-  declare export type FlipMoveProps = CommonProps & DelegatedProps & {
+  declare type PolymorphicProps = {
     duration: string | number,
     delay: string | number,
     staggerDurationBy: string | number,
     staggerDelayBy: string | number,
     enterAnimation: AnimationProp,
-    leaveAnimation: AnimationProp,
-    appearAnimation?: AnimationProp,
-    disableAnimations?: boolean, // deprecated, use disableAllAnimations instead
+    leaveAnimation: AnimationProp
   };
 
-  declare export default ComponentType<FlipMoveProps>;
+  declare type Hooks = {
+    onStart?: ChildHook,
+    onFinish?: ChildHook,
+    onStartAll?: ChildrenHook,
+    onFinishAll?: ChildrenHook
+  };
+
+  declare export type DelegatedProps = {
+    style?: ReactStyles
+  };
+
+  declare export type FlipMoveDefaultProps = BaseProps & PolymorphicProps;
+
+  declare export type CommonProps = BaseProps &
+    Hooks & {
+      children?: ChildrenArray<Child>
+    };
+
+  declare export type FlipMoveProps = FlipMoveDefaultProps &
+    CommonProps &
+    DelegatedProps & {
+      appearAnimation?: AnimationProp,
+      disableAnimations?: boolean // deprecated, use disableAllAnimations instead
+    };
```

Comment:
How does flow-typed work in terms of Flow versions? 

I believe in recent releases of Flow, this could be rewritten as:

```js
declare export type FlipMoveProps = {
  ...FlipMoveDefaultProps,
  ...CommonProps,
  ...DelegatedProps,
  appearAnimation?: AnimationProp,
  disableAnimations?: boolean
}
```

But yeah, maybe `flow-typed` aims to be compatible with earlier versions of Flow?

Reply:
You just specify version range. Thanks for the hint, I'll try that

Reply:
If I try to make this replacement, my `flow-typed` [tests](https://github.com/Hypnosphi/flow-typed/blob/master/definitions/npm/react-flip-move_v2.9.x/test_react-flip-move_v2.9.x.js) fail with following (when testing against Flow `v0.54`):
```
* react-flip-move_v2.9.x/flow_v0.53.0- (flow-v0.54.0): Unexpected Flow errors(2):
   Error: 0-test_react-flip-move_v2.9.x.js:19
                       v------------------------
    19:   getPosition={(node: HTMLElement) => ({
    20:     top: 0,
    21:     right: 0,
   ...:
    26:   })}
          -^ function. Could not decide which case to select
                                              v
    95:   declare export type FlipMoveProps = {
    96:     ...FlipMoveDefaultProps,
    97:     ...CommonProps,
   ...:
   101:   };
          ^ object type. See lib: react-flip-move_v2.9.x.js:95
     Case 1 may work:
      62:     getPosition: GetPosition,
                           ^^^^^^^^^^^ GetPosition. See lib: react-flip-move_v2.9.x.js:62
     But if it doesn't, case 2 looks promising too:
                                                v
      95:   declare export type FlipMoveProps = {
      96:     ...FlipMoveDefaultProps,
      97:     ...CommonProps,
     ...:
     101:   };
            ^ object type. See lib: react-flip-move_v2.9.x.js:95
     Please provide additional annotation(s) to determine whether case 1 works (or consider merging it with case 2):
                                                 v
      19:   getPosition={(node: HTMLElement) => ({
      20:     top: 0,
      21:     right: 0,
     ...:
      26:   })}
            ^ return

```

Plus it doesn't work in `v0.53`, so If I do that, I'd have to leave a separate libdef for one minor version only

Reply:
Yeah, totally fair. Wasn't aware that there was a difference! Thanks for trying it :)